File: services/web/src/main/java/com/example/web/converter/SyntheticSharesGenerator.java
package com.example.web.converter;

import com.example.gateway.api.model.ActionGW;
import com.example.gateway.api.model.ShareJournalGWModel;
import com.example.gateway.api.model.TransactionTypeGW;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

public class SyntheticSharesGenerator implements Function<List<ShareJournalGWModel>, List<ShareJournalGWModel>> {

    @Override
    public List<ShareJournalGWModel> apply(List<ShareJournalGWModel> shareJournalGWModels) {
        Map<String, ShareAggregator> stocks = new HashMap<>();

        shareJournalGWModels.stream()
                .filter(f -> f.getType().equals(TransactionTypeGW.SHARE))
                .forEach(s -> {
                    int quantity;
                    if (ActionGW.BUY == s.getAction()) {
                        quantity = s.getQuantity();
                    } else {
                        quantity = s.getQuantity() * -1;
                    }

                    ShareAggregator aggregator;
                    if (stocks.containsKey(s.getSymbol())) {
                        aggregator = stocks.get(s.getSymbol());
                    } else {
                        aggregator = new ShareAggregator(s.getSymbol());
                    }

                    aggregator.addQuantityAndPrice(quantity, s.getPrice());
                    stocks.put(s.getSymbol(), aggregator);
        });

        List<ShareJournalGWModel> synthetics = new ArrayList<>();
        stocks.forEach( (s, q) -> {
            if (q.getQuantity() != 0) {
                synthetics.add(ShareJournalGWModel.builder()
                        .withSymbol(s)
                        .withQuantity(q.getQuantity())
                        .withDate(OffsetDateTime.now().plusYears(1))
                        .withPrice(q.getAverageBoughtPrice())
                        .withAction(ActionGW.SELL)
                        .withType(TransactionTypeGW.SHARE)
                        .build());
            }
        });

        return synthetics;
    }
}

File: services/gateway/gateway-api/src/main/kotlin/com/example/gateway/api/core/util/SyntheticSharesGenerator.kt
package com.example.gateway.api.core.util

import com.example.common.converter.TimeConverter
import com.example.gateway.api.core.model.ActionType
import com.example.gateway.api.core.model.ShareAggregator
import com.example.gateway.api.core.model.ShareJournalModel
import com.example.gateway.api.core.model.TransactionType
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.math.abs

fun createSynthetic(models: List<ShareJournalModel>): List<ShareJournalModel> {
    val stockMap = HashMap<String, ShareAggregator>()

    models.stream()
            .filter { it.transactionType == TransactionType.SHARE && it.groupSelected && !it.legClosed}
            .forEach {
                val quantity = if (ActionType.BUY === it.action) {
                    it.quantity
                } else {
                    it.quantity * -1
                }

                val aggregator = if (stockMap.containsKey(it.symbol)) {
                    stockMap[it.symbol]!!
                } else {
                    ShareAggregator(symbol = it.symbol,
                            actualPrice = it.actualPrice,
                            preferredPrice = it.preferredPrice)
                }
                aggregator.addQuantityAndPrice(quantity, it.price)
                stockMap[it.symbol] = aggregator
            }

    val synthetics = ArrayList<ShareJournalModel>()
    stockMap.forEach {(symbol, aggregator) ->
        if (aggregator.outstandingShares != 0) {
            val averageBoughPrice = aggregator.averageBoughtPrice
            var syntheticActionType = ActionType.SELL
            if (aggregator.outstandingShares < 0) {
                syntheticActionType = ActionType.BUY
            }

            synthetics.add(ShareJournalModel(
                    transactionId = UUID.randomUUID(),
                    accountId = UUID.randomUUID(),
                    portfolioId = UUID.randomUUID(),
                    symbol = symbol,
                    date = TimeConverter.nextYear(),
                    price = averageBoughPrice,
                    preferredPrice = aggregator.preferredPrice,
                    actualPrice = aggregator.actualPrice,
                    quantity = abs(aggregator.outstandingShares),
                    brokerFees = 0.0,
                    groupSelected = true,
                    legClosed = true,
                    transactionType = TransactionType.SYNTHETIC_SHARE,
                    action = syntheticActionType
            ))

        }
    }

    return synthetics
}